xds/src/main/java/io/grpc/xds/LrsLoadBalancer.java

```diff
@@ -0,0 +1,121 @@
+import io.grpc.xds.EnvoyProtoData.Locality;
+import io.grpc.xds.LrsLoadBalancerProvider.LrsConfig;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.Objects;
+
+/**
+ * Load balancer for lrs policy.
+ */
+final class LrsLoadBalancer extends LoadBalancer {
+  private final LoadBalancer.Helper helper;
+  private GracefulSwitchLoadBalancer switchingLoadBalancer;
+  private LoadStatsStore loadStatsStore;
+  private String clusterName;
+  private String edsServiceName;
+  private Locality locality;
+  private String childPolicyName;
+
+  LrsLoadBalancer(LoadBalancer.Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    LrsConfig config = (LrsConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+    LoadStatsStore store =
+        resolvedAddresses.getAttributes().get(XdsAttributes.ATTR_CLUSTER_SERVICE_LOAD_STATS_STORE);
+    checkAndSetUp(config, store);
+
+    if (switchingLoadBalancer == null) {
+      loadStatsStore.addLocality(config.locality);
+      final ClientLoadCounter counter = loadStatsStore.getLocalityCounter(config.locality);
+      LoadBalancer.Helper loadRecordingHelper = new ForwardingLoadBalancerHelper() {
+        @Override
+        protected Helper delegate() {
+          return helper;
+        }
+
+        @Override
+        public void updateBalancingState(ConnectivityState newState, SubchannelPicker newPicker) {
+          @SuppressLint("VisibleForTests")
+          SubchannelPicker loadRecordingPicker =
+              new LoadRecordingSubchannelPicker(counter, newPicker);
+          super.updateBalancingState(newState, loadRecordingPicker);
+        }
+      };
+      switchingLoadBalancer = new GracefulSwitchLoadBalancer(loadRecordingHelper);
+    }
+    String updatedChildPolicyName = config.childPolicy.getProvider().getPolicyName();
+    if (!Objects.equals(childPolicyName, updatedChildPolicyName)) {
+      switchingLoadBalancer.switchTo(config.childPolicy.getProvider());
+      childPolicyName = updatedChildPolicyName;
+    }
+    ResolvedAddresses downStreamResult =
+        resolvedAddresses.toBuilder()
+            .setLoadBalancingPolicyConfig(config.childPolicy.getConfig())
+            .build();
+    switchingLoadBalancer.handleResolvedAddresses(downStreamResult);
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    if (switchingLoadBalancer != null) {
+      switchingLoadBalancer.handleNameResolutionError(error);
+    } else {
+      helper.updateBalancingState(ConnectivityState.TRANSIENT_FAILURE, new ErrorPicker(error));
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    if (switchingLoadBalancer != null) {
+      loadStatsStore.removeLocality(locality);
+      switchingLoadBalancer.shutdown();
+    }
+  }
+
+  private void checkAndSetUp(LrsConfig config, LoadStatsStore store) {
```

Comment:
Do you want to check if `store` is null?

Reply:
`checkState(
        loadStatsStore == null || loadStatsStore.equals(store),
        "loadStatsStore should not change");`
should have covered it.

Reply:
Are you allowing the `store` argument to be nullable?